compose/config/types.py

```diff
@@ -118,58 +119,96 @@ def normalize_paths_for_engine(external_path, internal_path):
+def splitdrive(path):
+    if len(path) == 0:
+        return ('', '')
+    if path[0] in ['.', '\\', '/', '~']:
+        return ('', path)
+    return ntpath.splitdrive(path)
 
 
 class VolumeSpec(namedtuple('_VolumeSpec', 'external internal mode')):
 
     @classmethod
-    def parse(cls, volume_config):
-        """Parse a volume_config path and split it into external:internal[:mode]
-        parts to be returned as a valid VolumeSpec.
-        """
-        if IS_WINDOWS_PLATFORM:
-            # relative paths in windows expand to include the drive, eg C:\
-            # so we join the first 2 parts back together to count as one
-            drive, tail = os.path.splitdrive(volume_config)
-            parts = tail.split(":")
-
-            if drive:
-                parts[0] = drive + parts[0]
-        else:
-            parts = volume_config.split(':')
+    def _parse_unix(cls, volume_config):
+        parts = volume_config.split(':')
 
         if len(parts) > 3:
             raise ConfigurationError(
                 "Volume %s has incorrect format, should be "
                 "external:internal[:mode]" % volume_config)
 
         if len(parts) == 1:
-            external, internal = normalize_paths_for_engine(
-                None,
-                os.path.normpath(parts[0]))
+            external = None
+            internal = os.path.normpath(parts[0])
         else:
-            external, internal = normalize_paths_for_engine(
-                os.path.normpath(parts[0]),
-                os.path.normpath(parts[1]))
+            external = os.path.normpath(parts[0])
+            internal = os.path.normpath(parts[1])
 
         mode = 'rw'
         if len(parts) == 3:
             mode = parts[2]
 
         return cls(external, internal, mode)
 
+    @classmethod
+    def _parse_win32(cls, volume_config):
+        # relative paths in windows expand to include the drive, eg C:\
+        # so we join the first 2 parts back together to count as one
+        mode = 'rw'
+
+        drive, tail = splitdrive(volume_config)
+        parts = tail.split(':', 1)
+        if drive:
+            parts[0] = drive + parts[0]
+        if len(parts) == 1:
+            external, internal = normalize_paths_for_engine(
+                None, os.path.normpath(parts[0])
+            )
+        else:
+            external = parts[0]
+            drive, tail = splitdrive(parts[1])
+            parts = tail.split(':')
+            if len(parts) > 2:
+                raise ConfigurationError(
+                    "Volume %s has incorrect format, should be "
+                    "external:internal[:mode]" % volume_config
+                )
+            if drive:
+                parts[0] = drive + parts[0]
+            external, internal = normalize_paths_for_engine(
+                os.path.normpath(external), os.path.normpath(parts[0])
+            )
+            if len(parts) > 1:
+                mode = parts[1]
```

Comment:
Looks like there's some repeated logic here where we call `splitdrive`, call `split(':')` on the tail, then prepend the drive onto the first part. Could it be factored out into a helper function?